mcstatus/motd.py

```diff
@@ -0,0 +1,820 @@
+                break
+
+            if item == next_item:
+                to_remove.add(index)
+
+        return to_remove
+
+    def remove_double_colors(self) -> set[int]:
+        """Get indices of all doubled color items.
+
+        As colors (obviously) override each other, we only ever care about the last one, ignore
+        the previous ones. (for example: specifying red color, then orange, then yellow, then some text
+        will just result in yellow text)
+        """
+        to_remove: set[int] = set()
+
+        prev_color: int | None = None
+        for index, item in enumerate(self.parsed):
+            if isinstance(item, (MinecraftColor, WebColor)):
+                # If we found a color after another, remove the previous color
+                if prev_color is not None:
+                    to_remove.add(prev_color)
+                prev_color = index
+
+            # If we find a string, that's what our color we found previously applies to,
+            # set prev_color to None, marking this color as used
+            if isinstance(item, str):
+                prev_color = None
+
+        return to_remove
+
+    def obtain_indices_of_all_unused_formatting_before_color(self) -> set[int]:
+        """Obtain indices of all unused formatting items before colors.
+
+        Colors override any formatting before them, meaning we only ever care about the color, and can
+        ignore all formatting before it. (For example: specifying bold formatting, then italic, then yellow,
+        will just result in yellow text.)
+        """
+        to_remove: set[int] = set()
+
+        for index, item in enumerate(self.parsed):
+            try:
+                next_item = self.parsed[index + 1]
+            except IndexError:  # Last item (without any next item)
+                break
+
+            if isinstance(item, Formatting) and isinstance(next_item, (MinecraftColor, WebColor)):
+                to_remove.add(index)
+
+        return to_remove
+
+    def get_indices_of_all_items_before_text_at_the_end(self) -> set[int]:
+        """Get indices of all items before text at the end.
+
+        Any color/formatting items only make sense when they apply to some text.
+        If there are some at the end, after the last text, they're pointless and
+        can be removed.
+        """
+        to_remove: set[int] = set()
+
+        for index, item in enumerate(self.parsed):
+            if index == len(self.parsed) - 1 and not isinstance(item, str):
+                to_remove.add(index)
+
+        return to_remove
+
+    def remove_empty_text(self) -> set[int]:
+        """Remove empty text values."""
+        to_remove: set[int] = set()
+
+        for index, item in enumerate(self.parsed):
+            if isinstance(item, str) and len(item) == 0:
+                to_remove.add(index)
+
+        return to_remove
+
+
+_HOOK_RETURN_TYPE = t.TypeVar("_HOOK_RETURN_TYPE")
+
+
+class BaseTransformer(abc.ABC, t.Generic[_HOOK_RETURN_TYPE]):
```

Comment:
Well at that point, `html` shouldn't be using a class at all, I understand that it may be complex as a single function, but there's a simple solution then: split the function into multiple private functions, each handling different parts of that complex logic.

I changed the transformers in the way I did to make their initialization truly meaningful (i.e. it now actually makes sense to create instances of transformer, as they potentially contain some transformer configuration, which is then used in the transform method), the original implementation didn't make any sense OOP wise, and at that point, class usage wasn't justified at all. As to how they work specifically to achieve this new structure, I just chose some logic that was simple enough yet still very extensible, if you want to change it, you can, but it needs to make sense and follow the same general idea, I will NOT accept a transformer class that doesn't even make sense as a class, at that point, just use functions.

Reply:
Fixed [here](https://github.com/py-mine/mcstatus/compare/a8a1a5b59a663740f5c91918bf88ac6aa1cfec27..e53b5dd85b9e267c18d8532707b02e8181b6337b).